src/lib/Imp.hs

```diff
@@ -258,6 +295,12 @@ toImpOp (maybeDest, op) = case op of
   RecordSplit  _ _ -> error "Unreachable: should have simplified away"
   VariantLift  _ _ -> error "Unreachable: should have simplified away"
   VariantSplit _ _ -> error "Unreachable: should have simplified away"
+  DataConTag con -> case con of
+    (Con (SumAsProd _ tag _)) -> returnVal tag
+    (DataCon _ _ i _) -> returnVal $ TagRepVal $ fromIntegral i
+    _ -> error $ "Not a data constructor: " ++ pprint con
+  ToEnum ~ty@(TypeCon (DataDef _ _ cons) _) i ->
+    returnVal $ Con $ SumAsProd ty i (map (const []) cons)
```

Comment:
We should check that the ADT doesn't have any fields in here

Reply:
Good idea. Added a check to Type.hs (we can throw exceptions properly there).